xds/src/main/java/com/linecorp/centraldogma/xds/internal/ControlPlanePlugin.java

```diff
@@ -0,0 +1,343 @@
+        for (Entry<?> entry : entries) {
+            try {
+                final ClusterLoadAssignment.Builder endpointBuilder = ClusterLoadAssignment.newBuilder();
+                JsonFormatUtil.parser().merge(entry.contentAsText(), endpointBuilder);
+                endpointsBuilder.add(endpointBuilder.build());
+            } catch (Throwable e) {
+                logger.warn("Unexpected exception is raised while building an endpoint using {}" +
+                            ". Try watching after {} seconds..", entry, BACKOFF_SECONDS, e);
+                return false;
+            }
+        }
+
+        setNewSnapshot(cache, ResourceType.ENDPOINT,
+                       CentralDogmaSnapshotResources.create(endpointsBuilder.build(), revision));
+        return true;
+    }
+
+    private boolean listeners(Collection<Entry<?>> entries, Revision revision, SimpleCache<String> cache) {
+        final Builder<Listener> listenersBuilder = ImmutableList.builder();
+        for (Entry<?> entry : entries) {
+            try {
+                final Listener.Builder listenerBuilder = Listener.newBuilder();
+                JsonFormatUtil.parser().merge(entry.contentAsText(), listenerBuilder);
+                listenersBuilder.add(listenerBuilder.build());
+            } catch (Throwable e) {
+                logger.warn("Unexpected exception is raised while building a listener using {}" +
+                            ". Try watching after {} seconds..", entry, BACKOFF_SECONDS, e);
+                return false;
+            }
+        }
+
+        setNewSnapshot(cache, ResourceType.LISTENER,
+                       CentralDogmaSnapshotResources.create(listenersBuilder.build(), revision));
+        return true;
+    }
+
+    private boolean routes(Collection<Entry<?>> entries, Revision revision, SimpleCache<String> cache) {
+        final Builder<RouteConfiguration> routesBuilder = ImmutableList.builder();
+        for (Entry<?> entry : entries) {
+            try {
+                final RouteConfiguration.Builder routeBuilder = RouteConfiguration.newBuilder();
+                JsonFormatUtil.parser().merge(entry.contentAsText(), routeBuilder);
+                routesBuilder.add(routeBuilder.build());
+            } catch (Throwable e) {
+                logger.warn("Unexpected exception is raised while building a route using {}" +
+                            ". Try watching after {} seconds..", entry, BACKOFF_SECONDS, e);
+                return false;
+            }
+        }
+
+        setNewSnapshot(cache, ResourceType.ROUTE,
+                       CentralDogmaSnapshotResources.create(routesBuilder.build(), revision));
+        return true;
+    }
+
+    @SuppressWarnings("unchecked")
+    private void setNewSnapshot(SimpleCache<String> cache, ResourceType resourceType,
+                                SnapshotResources<?> resources) {
+        snapshotLock.lock();
```

Comment:
Q) Do multiple threads access this method?

Reply:
It was, but it isn't anymore. Thanks!